src/hotspot/share/gc/z/zRelocate.cpp

```diff
@@ -56,24 +56,26 @@ class ZRelocateRootsIteratorClosure : public ZRootsIteratorClosure {
 
 class ZRelocateRootsTask : public ZTask {
 private:
-  ZRootsIterator                _roots;
   ZRelocateRootsIteratorClosure _cl;
 
 public:
   ZRelocateRootsTask() :
-      ZTask("ZRelocateRootsTask"),
-      _roots(true /* visit_jvmti_weak_export */) {}
+      ZTask("ZRelocateRootsTask") {}
 
   virtual void work() {
+    // Allocation path assumes that relocating GC threads are ZWorkers
+    assert(ZThread::is_worker(), "Relocation code needs to be run as a worker");
+    assert(ZThread::worker_id() == 0, "No multi-thread support");
+
     // During relocation we need to visit the JVMTI
     // export weak roots to rehash the JVMTI tag map
-    _roots.oops_do(&_cl);
+    ZRelocateRoots::oops_do(&_cl);
```

Comment:
I would have been more explicit here, something like this:

```
class ZRelocateJVMTIWeakExportTask : public ZTask {
public:
  ZRelocateJVMTIWeakExportTask() :
      ZTask("ZRelocateJVMTIWeakExportTask") {}

  virtual void work() {
    AlwaysTrueClosure always_alive;
    ZRelocateRootsIteratorClosure _cl;
    JvmtiExport::weak_oops_do(&always_alive, &cl);
  }
};

void ZRelocate::start() {
  // During relocation we need to visit the JVMTI
  // export weak roots to rehash the JVMTI tag map
  ZStatTimer timer(ZSubPhasePauseRelocateJVMTIWeakExport);
  ZRelocateJVMTIWeakExportTask task;
  _workers->run_serial(&task);
}
```

But this code will go away soon anyway, so you decide.

Reply:
I agree, and I actually started that way, but hit some minor snags that ZSubPhasePauseRelocateJVMTIWeakExport wasn't available to zRelocate.cpp, and I wasn't sure about moving a *PhasePause* instance out from the zRootsIterator.cpp.